strutil/quantity/quantity.go

```diff
@@ -200,3 +201,262 @@ func FormatDuration(dt float64) string {
+// In the Compact case, only the first two places will be rendered,
+// discarding the rest. This allows for a more predictable width (at the
+// expense of accuracy).
+type Places uint32
+
+const (
+	ShowVerbose Places = 8
+	ShowCompact Places = 2
+)
+
+// Space delimit options
+type SpaceMode bool
+
+const (
+	SpaceOff SpaceMode = false
+	SpaceOn  SpaceMode = true
+)
+
+// The minimum Duration parameter allows the rendering to discard the units
+// of time less than the minimum. However, depending on what we are
+// rendering, we either need to peform a ceiling, floor or rounding
+// operation with the remainder below the minimum.
+type RenderMode uint32
+
+const (
+	TimeLeft RenderMode = iota
+	TimePassed
+	TimeRounded
+)
+
+// FormatDurationGeneric formats time duration similar to systemd
+// format_timespan, but with options allowing for a more compact
+// arrangement, including shortened suffixes. The function is provided
+// with dt in seconds (as in the output of time.Now().Seconds()). The 'min'
+// and 'max' Duration allows limiting the unit range rendered. Place values
+// less than 'min' will be processed according to RenderMode. RenderMode
+// allows rounding, flooring or ceiling to be applied to place value below
+// 'min'. Durations above the 'max' place value will result in "ages!".
+func FormatDurationGeneric(dt float64, min Duration, max Duration, count Places, mode RenderMode, space SpaceMode) string {
+	var units = map[Duration]string{
+		Year:    "y",
+		Day:     "d",
+		Hour:    "h",
+		Minute:  "m",
+		Second:  "s",
+		MSecond: "ms",
+		USecond: "µs",
+		NSecond: "ns",
+	}
+
+	// We need to access the map in order using an ordered array
+	// Sort into iteration order: y, d, h, m, s, ...
+	var keys []Duration
+	for k := range units {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] > keys[j] })
+
+	var render string
+
+	// Invalid case: min > max
+	if min > max {
+		return "inv!"
+	}
+
+	// Special case: zero duration
+	if dt <= 0 {
+		return "0" + units[min]
+	}
+
+	// Special case: render as "ages!"
+	if dt >= float64(math.MaxUint64) {
+		return "ages!"
+	}
+
+	// Fractional seconds to nanoseconds
+	delta := Duration(dt * float64(Second))
+
+	// If we only render a subset of place values, we need to apply
+	// the render mode to the last place value. We do this indirectly
+	// by moving up the 'min' accuracy.
+	if count == ShowCompact {
+		for i, key := range keys {
+
+			// The duration is smaller than this place value, skip
+			if delta < key {
+				continue
+			}
+
+			// Compact mode only renders two places, up until seconds
+			// The indexing is safe as places beyond seconds exist
+			if key >= Minute {
+				index := i + 1
+				if keys[index] > min {
+					min = keys[index]
+				}
+				break
+			}
+
+			// Only one place if we are s, ms, us or ns
+			if key < Minute {
+				if key > min {
+					min = key
+				}
+				break
+			}
+		}
```

Comment:
Now that it has been simplified and I better understand what it's doing, it seems to me that it could be simplified further:
```suggestion
		var mainUnit Duration
		for i, key := range keys {
			if delta >= key {
				if key >= Minute {
					// Compact mode only renders two places, up until seconds
					// The indexing is safe as places beyond seconds exist
					mainUnit = keys[i + 1]
				} else {
					mainUnit = key
				}
				break
			}
		}

		if mainUnit > min {
			min = mainUnit
		}
```
Do you agree that these two are equivalent?

Reply:
Perfect. Fixed.